datasets/ccagt/ccagt.py

```diff
@@ -0,0 +1,318 @@
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _download_and_extract_all(self, dl_manager):
+        def extracted_by_slide(paths):
+            return {get_slide_id(path): dl_manager.extract(path) for path in paths}
+
+        data_dir = dl_manager.download_and_extract(_DATA_URL)
+        base_path = os.path.join(data_dir, "wg4bpm33hj-2")
+
+        logger.info("Extracting images...")
+        self.images_base_dir = os.path.join(base_path, "images")
+        images_to_extract = [
+            os.path.join(self.images_base_dir, fn) for fn in os.listdir(self.images_base_dir) if fn.endswith(".zip")
+        ]
+        self.images_extracted = extracted_by_slide(images_to_extract)
+
+        if self.config.name == "semantic_segmentation":
+            logger.info("Extracting masks...")
+            self.masks_base_dir = os.path.join(base_path, "masks")
+            masks_to_extract = [
+                os.path.join(self.masks_base_dir, fn) for fn in os.listdir(self.masks_base_dir) if fn.endswith(".zip")
+            ]
+            self.masks_extracted = extracted_by_slide(masks_to_extract)
+        elif self.config.name == "object_detection":
+            logger.info("Reading COCO OD file...")
+            ccagt_OD_COCO_path = os.path.join(base_path, "CCAgT_COCO_OD.json")
+            with open(ccagt_OD_COCO_path, "r", encoding="utf-8") as json_file:
+                coco_OD = json.load(json_file)
+
+            self._imageid_to_coco_OD_annotations = defaultdict(list)
+            for labels in coco_OD["annotations"]:
+                self._imageid_to_coco_OD_annotations[labels["image_id"]].append(labels)
+
+        logger.info("Loading dataset info...")
+        ccagt_raw_path = os.path.join(base_path, "CCAgT.parquet.gzip")
+        with open(ccagt_raw_path, "rb") as f:
+            self._ccagt_info = pd.read_parquet(f, columns=["image_name", "category_id", "image_id", "slide_id"])
+        self._bn_to_imageid = pd.Series(
+            self._ccagt_info["image_id"].values, index=self._ccagt_info["image_name"]
+        ).to_dict()
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators."""
+
+        def build_path(basename, tp="images"):
+            slide = basename.split("_")[0]
+            if tp == "images":
+                dir_path = self.images_extracted[slide]
+                ext = ".jpg"
+            else:
+                dir_path = self.masks_extracted[slide]
+                ext = ".png"
+
+            return os.path.join(dir_path, slide, basename + ext)
+
+        def images_and_masks(basenames):
+            for bn in basenames:
+                yield build_path(bn), build_path(bn, "masks")
+
+        def images_and_boxes(basenames):
+            for bn in basenames:
+                image_id = self._bn_to_imageid[bn]
+                labels = [
+                    {"bbox": annotation["bbox"], "label": annotation["category_id"]}
```

Comment:
To respect the `0, ..., num_classes-1` label range

```suggestion
                    {"bbox": annotation["bbox"], "label": annotation["category_id"] - 1}
```